lib/ascii_binder/engine.rb

```diff
@@ -0,0 +1,872 @@
+      end
+      config_data.each do |topic_group|
+        validate_topic_group(topic_group, { :path => topic_group['Dir'] })
+        # Now buzz through the topics
+        topic_group['Topics'].each do |topic|
+          # Is this an actual topic or a subtopic group?
+          is_subtopic_group = topic.has_key?('Dir') and topic.has_key?('Topics') and not topic.has_key?('File')
+          is_topic_item = topic.has_key?('File') and not topic.has_key?('Dir') and not topic.has_key?('Topics')
+          if not is_subtopic_group and not is_topic_item
+            raise "This topic could not definitively be determined to be a topic item or a subtopic group:\n#{topic.inspect}"
+          end
+          if is_topic_item
+            validate_topic_item(topic, { :group => topic_group['Name'], :group_id => topic_group['ID'], :path => topic_group['Dir'] })
+          elsif is_subtopic_group
+            topic_path = "#{topic_group['Dir']}/#{topic['Dir']}"
+            validate_topic_group(topic, { :path => topic_path, :parent_id => topic_group['ID'] })
+            topic['Topics'].each do |subtopic|
+              validate_topic_item(subtopic, { :group => "#{topic_group['Name']}/#{topic['Name']}", :group_id => topic['ID'], :path => topic_path })
+            end
+          end
+        end
+      end
+      config_data
+    end
+
+    def camelize text
+      text.gsub(/[^0-9a-zA-Z ]/i, '').split(' ').map{ |t| t.capitalize }.join
+    end
+
+    def nav_tree distro, branch_build_config
+      navigation = []
+      branch_build_config.each do |topic_group|
+        next if not topic_group['Distros'].include?(distro.id)
+        next if topic_group['Topics'].select{ |t| t['Distros'].include?(distro.id) }.length == 0
+        topic_list = []
+        topic_group['Topics'].each do |topic|
+          next if not topic['Distros'].include?(distro.id)
+          if topic.has_key?('File')
+            topic_list << {
+              :path => "../#{topic_group['Dir']}/#{topic['File']}.html",
+              :name => topic['Name'],
+              :id   => topic['ID'],
+            }
+          elsif topic.has_key?('Dir')
+            next if topic['Topics'].select{ |t| t['Distros'].include?(distro.id) }.length == 0
+            subtopic_list = []
+            topic['Topics'].each do |subtopic|
+              next if not subtopic['Distros'].include?(distro.id)
+              subtopic_list << {
+                :path => "../#{topic_group['Dir']}/#{topic['Dir']}/#{subtopic['File']}.html",
+                :name => subtopic['Name'],
+                :id   => subtopic['ID'],
+              }
+            end
+            topic_list << { :name => topic['Name'], :id => topic['ID'], :topics => subtopic_list }
+          end
+        end
+        navigation << { :name => topic_group['Name'], :id => topic_group['ID'], :topics => topic_list }
+      end
+      navigation
+    end
+
+    def asciidoctor_page_attrs(more_attrs=[])
+      [
+        'source-highlighter=coderay',
```

Comment:
This is really part of a broader discussion around decoupling AsciiBinder and Asciidoctor, so that the converter used by AsciiBinder is totally configurable. Right now our invocation of Asciidoctor is [hard-coded](https://github.com/redhataccess/ascii_binder/blob/master/lib/ascii_binder/helpers.rb#L714) and so we need to replace these direct calls to a ruby object with system calls to a user-specifiable parser with user-specified runtime arguments.